src/big_int10.cpp

```
@@ -0,0 +1,562 @@
+    return *this;
+}
+
+BigInt10 BigInt10::operator%(const BigInt10 &b) const
+{
+    /// c=a%b;
+    BigInt10 c(*this);
+    c %= b;
+    return c;
+}
+
+BigInt10 &BigInt10::operator^=(const BigInt10 &b)
+{
+    // this^=b;
+    *this = power(*this, b);
+    return *this;
+}
+
+BigInt10 BigInt10::operator^(const BigInt10 &b) const
+{
+    // c=this^b
+    BigInt10 c(*this);
+    c ^= b;
+    return c;
+}
+
+std::string BigInt10::toString() const
+{
+    std::ostringstream oss;
+    oss << *this;
+    return oss.str();
+}
+
+uint64_t BigInt10::toU64() const
+{
+    BigInt10 base = MAX_VAL_64;
+    if (base.isSmallerThanUnsigned(*this))
+        throw std::invalid_argument("this bigger than uint64");
+    uint64_t result = 0;
+    for (int i = limbsCount() - 1; i >= 0; i--) {
+        result *= 10;
+        result += limbs[i];
+    }
+    return result;
+}
+
+void BigInt10::removeLeadingZeros()
+{
+    while (limbs.size() > 1 && limbs.back() == 0) {
+        limbs.pop_back();
+    }
+}
+
+void BigInt10::insertZroes(int n)
+{
+    limbs.insert(limbs.begin(), n, 0);
+}
+
+bool BigInt10::isSmallerThanUnsigned(const BigInt10 &b) const
+{
+    //  ans=a<b
+    if (limbsCount() > b.limbsCount())
+        // a is longer
+        return false;
+    if (limbsCount() < b.limbsCount())
+        // b is longer
+        return true;
+    // same length
+    for (int i = limbsCount() - 1; i >= 0; i--)
+        if (limbs[i] != b.limbs[i])
+            return limbs[i] < b.limbs[i];
+    // they are equal
+    return false;
+}
+
+void BigInt10::prefixPlusPlusUnsigned()
+{
+    //++a
+    int i, n = limbsCount();
+    // zero alll the MAX_VAL lsb's
+    for (i = 0; i < n && limbs[i] == MAX_VAL_10; i++) {
+        limbs[i] = 0;
+    }
+    if (i == n)
+        // it was all MAX_VAL
+        limbs.push_back(1);
+    else
+        limbs[i]++;
+}
+
+void BigInt10::prefixMinusMinusUnsigned()
+{
+    //--a
+    if (isZero())
+        throw std::invalid_argument("cant --0");
```

Should return an error code?

---

this is internal class not exposed to the user, and actions on that class are wrapped with try catch